In [2]:
import numpy as np
import pandas as pd
import os

In [11]:
print(os.getcwd())

/workspaces/ML-Zoomcamp-Homework/01-intro


Q1. Pandas version

In [4]:
pd.__version__

'3.0.2'

In [3]:
df = pd.read_csv("../car_fuel_efficiency.csv")

Q2. Records count

In [14]:
df.index

RangeIndex(start=0, stop=9704, step=1)

The answer is 9704

Q3. Fuel types

In [15]:
df.head(5)

,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,model_year,origin,fuel_type,drivetrain,num_doors,fuel_efficiency_mpg
0,170,3.0,159.0,3413.433759,17.7,2003,Europe,Gasoline,All-wheel drive,0.0,13.231729
1,130,5.0,97.0,3149.664934,17.8,2007,USA,Gasoline,Front-wheel drive,0.0,13.688217
2,170,NaN,78.0,3079.038997,15.1,2018,Europe,Gasoline,Front-wheel drive,0.0,14.246341
3,220,4.0,NaN,2542.392402,20.2,2009,USA,Diesel,All-wheel drive,2.0,16.912736
4,210,1.0,140.0,3460.870990,14.4,2009,Europe,Gasoline,All-wheel drive,2.0,12.488369


In [16]:
df.fuel_type.nunique()

2

The answer is 2

Q4. Missing values

In [7]:
missing_per_column = df.isnull().sum()
cols_with_missing = (missing_per_column > 0).sum()
print(cols_with_missing)

4


Answer is 4

Question 5

In [8]:
sorted_df = df.groupby("origin")["fuel_efficiency_mpg"].max()
print(sorted_df)

origin
Asia      23.759123
Europe    25.967222
USA       24.971452
Name: fuel_efficiency_mpg, dtype: float64


Answer is 23.75

Question 6

In [9]:
df["horsepower"].median()

np.float64(149.0)

In [24]:
freq_highest = df["horsepower"].mode()[0]
print(freq_highest)

152.0


In [25]:
df_filled = df.horsepower.fillna(freq_highest)
print(df_filled)

0       159.0
1        97.0
2        78.0
3       152.0
4       140.0
        ...  
9699    164.0
9700    154.0
9701    138.0
9702    177.0
9703    140.0
Name: horsepower, Length: 9704, dtype: float64


In [27]:
df_filled.median()

np.float64(152.0)

Answer: from 149 to 152, the median has increased

Question 7

In [ ]:
df[df["origin"] == 'Asia'][["vehicle_weight", "model_year"]]

,vehicle_weight,model_year
8,2714.219310,2016
12,2783.868974,2010
14,3582.687368,2007
20,2231.808142,2011
21,2659.431451,2016
...,...,...
9688,3948.404625,2018
9692,3680.341381,2016
9693,2545.070139,2012
9698,3107.427820,2005
